src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -294,7 +291,90 @@ public void onFailure(@NonNull Throwable t) {
                     e.afterStepExecutionsResumed();
                 }
 
-            }, Timer.get()); // We always hold RunMap and WorkflowRun locks here, so we resume steps on a different thread to avoid potential deadlocks. See JENKINS-67351.
+            }, MoreExecutors.directExecutor());
         }
     }
+
+    /** Calls {@link StepExecution#onResume} for each step in a running build.
+     * Does so in parallel, but always completing enclosing blocks before the enclosed step.
+     * A simplified version of https://stackoverflow.com/a/67449067/12916, since this should be a tree not a general DAG.
+     */
+    private static final class ParallelResumer {
```

Comment:
Is this for correctness or an optimization?

Reply:
It is an optimization. Previously all steps were resumed serially in a topological order. (Or I hope `getCurrentExecutions` enforced a topological order; the sorting by `CpsThread` is a bit opaque to me.) The problem in https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/180 arises when you have a bunch of parallel branches all running `node` and all of them are going to fail due to missing agents. Previously, Jenkins would wait 5m for the first agent, report that it was gone, then wait 5m for the next agent, etc. Now all the `node` blocks are resumed in parallel, followed by the `sh` steps inside them, etc. Have a test demonstrating this in `workflow-durable-task-step` but only now got an incremental build from https://github.com/jenkinsci/workflow-cps-plugin/pull/534 that I needed.

Reply:
Ok, I guess it's kind of unusual for `ExecutorStepExecution.onResume` to block if no other steps do so, but looking through https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/180 it seems like you already tried various alternatives, so this seems fine.

Reply:
Basically this is the replacement for the `TryRepeatedly` pickle. I played with a lot of alternatives indeed. I am not claiming this approach is ideal but it seems to be straightforward enough and do the job.

Earlier, before hitting on the idea of having `onResume` block, I had hoped to arrange it so that the program would resume right away, and then a `sh` step running on the dead agent would eventually figure out there was no hope and abort. I ran into problems in functional tests, though, things like

```groovy
node('agent') {
  input 'Proceed?' // restart here
  sh 'true'
}
```

The program would run past `input` to the `sh` step, which would then either fail with a `MissingContextVariableException` (confusing) or block the CPS VM thread inside `DSL` (which will throw an error if it blocks >5s). Admittedly this is an artificial case (you should not hold an executor like that) but a lot of tests were failing that I had to work around in artificial ways and it seemed uncomfortable. It would have been possible to fix all these cases but only in pretty intrusive ways: by defining a new `StepContext` object in lieu of `FilePath` (perhaps `DynamicFilePathContext.Representation`) which could be deserialized without pickles, but every step currently expecting `FilePath` would need to be adjusted to expect this instead, and then the step would need to have complicated logic to wait for the new contextual object to be ready (translatable to a live `FilePath`) before doing anything useful. 

The compromise I settled on is closer to the original pickle behavior in that CPS code and steps do not run until an attempt is made to restore the original context: until all open `node` blocks have gotten their agent to reconnect, or timed out trying. The difference is only that in case this fails (with a timeout), the error is thrown out of `StepExecution.onResume` and thus `StepContext.onFailure` so it becomes a properly modeled `Throwable` with a CPS VM stack trace that we can catch and handle—unlike the original situation where the entire build would be effectively hard-killed with no possible cleanup.